### PR TITLE
STITCH-2736: adds various improvements to performance testing harness

### PR DESCRIPTION
**Pull request by tkaye407**

### CODE REVIEW

**Comment by tkaye407** [comment]
This fixed a few problems including: 
1. Fixed the network metrics collection which had the same big found in swift of only working on the first iteration
2. Adopted the failure protocol we discussed as a group
3. Takes all test params from local.properties (and passes these properties in .evg.yml) or uses separate set of defaults
4. Fixed some strange teardown / setup bugs that were causing problems on local tests
5. Fixed some teardown / setup bugs that were cuasing problems on prod tests
6. Separate out SyncPerformanceTestContext into LocalPerformanceTestContext and ProductionPerformanceTestContext

What is still needed: 
1. Disk usage is still off and I would like to continue to look into that in a different ticket 
2. We still cannot get CPU Usage

File: android/services/mongodb-remote-perftests/build.gradle [hunk 1]
```diff
@@ -31,11 +31,35 @@ android {
         if (!properties.getProperty("test.stitch.androidPerfStitchAPIKey", "").isEmpty()) {
             testInstrumentationRunnerArgument "test.stitch.androidPerfStitchAPIKey", properties.getProperty("test.stitch.androidPerfStitchAPIKey", "")
         }
-        if (!properties.getProperty("test.stitch.androidPerfStitchHostname", "").isEmpty()) {
-            testInstrumentationRunnerArgument "test.stitch.androidPerfStitchHostname", properties.getProperty("test.stitch.androidPerfStitchHostname", "")
+        if (!properties.getProperty("test.stitch.perf.stitchHost", "").isEmpty()) {
```
Comment (adamchel) [inline android/services/mongodb-remote-perftests/build.gradle]: [nit] can we make an array of strings and do these in a for loop so it's a bit cleaner?

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceIntTestsHarness.kt [hunk 2]
```diff
@@ -125,74 +124,151 @@ class SyncPerformanceIntTestsHarness : BaseStitchAndroidIntTest() {
...
+                val updateDocument = Document("\$push", Document("results", runResultBson))
+                Tasks.await(outputColl.updateOne(filterDocument, updateDocument))
+            }
+
+            return true
+        }
+    }
+
+    private fun getPerformanceTestingContext(testParams: TestParams): SyncPerformanceTestContext {
+        if (testParams.stitchHostName == "https://stitch.mongodb.com") {
```
Comment (adamchel) [inline android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceIntTestsHarness.kt]: [nit] extract this out to a constant somewhere called `PROD_STITCH_HOSTNAME`, and put it maybe in `SyncPerformanceTestUtils`

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTestContext.kt [hunk 3]
```diff
@@ -175,37 +91,41 @@ class SyncPerformanceTestContext(
...
         partialResult.timeTaken = (Date().time - timeBefore).toDouble()
-        val statsAfter = StatFs(Environment.getExternalStorageDirectory().absolutePath)
-        val memFreeAfter = statsAfter.freeBlocksLong * statsAfter.blockSizeLong
-        partialResult.diskUsage = (memFreeBefore - memFreeAfter).toDouble()
+        val memFreeAfterData = StatFs(Environment.getDataDirectory().path).freeBytes
+        val memFreeAfterReal = StatFs("/data/user/0/com.mongodb.stitch.android.services.mongodb.performance.test").freeBytes
+        val memInUseData = StatFs(Environment.getDataDirectory().path).totalBytes - StatFs(Environment.getDataDirectory().path).freeBytes
+        partialResult.diskUsage = (memFreeBeforeData - memFreeAfterData).toDouble()
+
+        /* Log.d("perfTests", String.format("Memory Change Data %d (total): %d", (memFreeBeforeData - memFreeAfterData), memInUseData))
```
Comment (adamchel) [inline android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTestContext.kt]: [nit] remove this comment or add a comment explaining why it's commented out

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTestContext.kt [hunk 4]
```diff
@@ -175,37 +91,41 @@ class SyncPerformanceTestContext(
...
-                }
-            }
-        }
+    protected fun clearLocalDB() {
+        val syncedIds = Tasks.await(testColl.sync().syncedIds)
+        Tasks.await(testColl.sync().desyncMany(*syncedIds.toTypedArray()))
+        LocalMongoDbService.clearallLocalDBs()
+    }
+
+    /*
```
Comment (adamchel) [inline android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTestContext.kt]: [nit] remove these comments or add a comment explaining why they're commented out

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTestContext.kt [hunk 5]
```diff
@@ -175,37 +91,41 @@ class SyncPerformanceTestContext(
...
-                break
-            } catch (e: Exception) {
-                if (i < maxRetries) {
-                    Log.i(TAG, "Error deleting all documents: ${e.localizedMessage}, retrying")
-                } else {
-                    Log.e(TAG, "Error deleting all documents: ${e.localizedMessage}, no more retries")
-                }
-            }
-        }
+    protected fun clearLocalDB() {
```
Comment (adamchel) [inline android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTestContext.kt]: [q] why is this a function here and not in the local-specific test context?

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTests.kt [hunk 6]
```diff
@@ -24,91 +25,89 @@ class SyncPerformanceTests {
...
-    @Before
-    fun setup() {
-        this.testHarness.setup()
-    }
-
-    @After
-    fun teardown() {
-        this.testHarness.teardown()
-    }
+    // private var testHarness = SyncPerformanceIntTestsHarness()
```
Comment (adamchel) [inline android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTests.kt]: [nit] remove this comment
[q] also why is this out of the class now?

**Comment by adamchel** [comment]
Also it looks like your commits aren't signed so you'll need to likely need to make a new branch and squash your commits before merging

File: .evg.yml [hunk 7]
```diff
@@ -147,6 +147,18 @@ functions:
           echo "test.stitch.fcmSenderId=${test_fcm_sender_id}" >> local.properties
           echo "test.stitch.fcmApiKey=${test_fcm_api_key}" >> local.properties
           echo "test.stitch.androidPerfStitchAPIKey=${perf_android_api_key}" >> local.properties
+
+          echo "test.stitch.perf.stitchHost=http://localhost:9090" >> local.properties
+          echo "test.stitch.perf.iters=5" >> local.properties
+          echo "test.stitch.perf.hostname=Evergreen" >> local.properties
+          echo "test.stitch.perf.docSizes=1024,2048,4096,8192" >> local.properties
```
[the remark this answers is not in the document]
Comment (tkaye407) [reply .evg.yml]: This is why I sent out that table with what we want to be the defaults for the two types of cases. I figured let's err on the side of not overwhelming Evergreen as we define the tests and if we want to change them it is very easy

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTests.kt (hunk 6 above)
Comment (tkaye407) [reply android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTests.kt]: This was the reason that the network transport only worked the first time. The references to the transport got very strange. We had to do the same thing in iOS. Regardless it makes sense for us to only have one harness.

File: .evg.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (dkaminsky) [reply .evg.yml]: I’m not worried so much about the sizes we use initially until we figure out the extrema where we fall over and start ticketing out remediation

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceIntTestsHarness.kt [hunk 8]
```diff
@@ -249,26 +284,43 @@ class PartialResult {
...
-            mapOf(
-                "numDocs" to BsonInt32(numDocs),
-                "docSize" to BsonInt32(docSize),
-                "timeMs" to DataBlock(runTimes.toDoubleArray(), numOutliers).toBson(),
-                "networkSentBytes" to DataBlock(networkSentBytes.toDoubleArray(), numOutliers).toBson(),
-                "networkReceivedBytes" to DataBlock(networkReceivedBytes.toDoubleArray(), numOutliers).toBson(),
-                "memoryBytes" to DataBlock(memoryUsages.toDoubleArray(), numOutliers).toBson(),
-                "diskBytes" to DataBlock(diskUsages.toDoubleArray(), numOutliers).toBson(),
-                "activeThreadCounts" to DataBlock(activeThreadCounts.toDoubleArray(), numOutliers).toBson()
+        if (failures.size < (SyncPerformanceTestUtils.getNumIters() + 1) / 2) {
```
Comment (dkaminsky) [inline android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceIntTestsHarness.kt]: I don't follow this logic, please explain?

Comment (tkaye407) [reply android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceIntTestsHarness.kt]: Just that we consider the test a failure (dont get its metrics) if it fails half of the iterations

File: android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTests.kt (hunk 6 above)
Comment (dkaminsky) [reply android/services/mongodb-remote-perftests/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/performance/SyncPerformanceTests.kt]: +1 remove comments

**Pull request merged**